pydomcfg/domzgr/zco.py

```diff
@@ -0,0 +1,218 @@
+        self._ppkth2 = ppkth2
+        self._ppacr2 = ppacr2
+
+        ds = self.init_ds()
+
+        # computing coeff. if needed
+        if self._ppkth * self._ppacr > 0.0:
+            self._compute_pp()
+
+        # compute z3 depths of vertical levels
+        for k in range(self._jpk):
+
+            if self._ppkth * self._ppacr == 0.0:
+                # uniform zco grid
+                suT = -self.sigma(k, "T")
+                suW = -self.sigma(k, "W")
+                s1T = s1W = s2T = s2W = 0.0
+                a1 = a3 = a4 = 0.0
+                a2 = self._pphmax
+            else:
+                # stretched zco grid
+                suT = -self.sigma(k + 1, "T")
+                suW = -self.sigma(k + 1, "W")
+                s1T = self._stretch_zco(-self.sigma(k, "T"))
+                s1W = self._stretch_zco(-self.sigma(k, "W"))
+                a1 = self._ppsur
+                a2 = self._ppa0 * (self._jpk - 1)
+                a3 = self._ppa1 * self._ppacr
+                if self._ldbletanh:
+                    s2T = self._stretch_zco(-self.sigma(k, "T"), self._ldbletanh)
+                    s2W = self._stretch_zco(-self.sigma(k, "W"), self._ldbletanh)
+                    a4 = self._ppa2 * self._ppacr2
+                else:
+                    s2T = s2W = a4 = 0.0
+
+            ds["z3T"][{"z": k}] = self.compute_z3(suT, s1T, a1, a2, a3, s2T, a4)
+            ds["z3W"][{"z": k}] = self.compute_z3(suW, s1W, a1, a2, a3, s2W, a4)
+
+        # force first w-level to be exactly at zero
+        ds["z3W"][{"z": 0}] = 0.0
+
+        # compute e3 scale factors
+        dsz = self.compute_e3(ds)
+
+        return dsz
+
+    # --------------------------------------------------------------------------
+    def _compute_pp(self):
+        """
+        Compute the coefficients for zco grid if requested.
+
+        Raises
+        ------
+        ValueError
+            If ldbletanh = True and parametrs are equal to 0.
+
+        """
+        if (
+            self._ppsur == self.pp_to_be_computed
+            or self._ppa0 == self.pp_to_be_computed
+            or self._ppa1 == self.pp_to_be_computed
+        ):
+
+            aa = self._ppdzmin - self._pphmax / float(self._jpk - 1)
+            bb = np.tanh((1 - self._ppkth) / self._ppacr)
+            cc = self._ppacr / float(self._jpk - 1)
+            dd = np.log(np.cosh((self._jpk - self._ppkth) / self._ppacr)) - np.log(
+                np.cosh((1 - self._ppkth) / self._ppacr)
+            )
+
+            self._ppa1 = aa / (bb - cc * dd)
+            self._ppa0 = self._ppdzmin - self._ppa1 * bb
+            self._ppsur = -self._ppa0 - self._ppa1 * self._ppacr * np.log(
+                np.cosh((1 - self._ppkth) / self._ppacr)
+            )
+
+            if self._ldbletanh and self._ppa2 * self._ppkth2 * self._ppacr2 == 0.0:
+                raise ValueError(
+                    "ppa2, ppkth2 and ppacr2 must > 0. " "when ldbletanh = True"
+                )
```

Comment:
Idea I forgot to add, no need to think about it now. I find this is hard to digest for a couple of reasons: 1. There's no return, so I really have to read the whole method to see what's going on, long statements, not sure if this is the right place for that error as it looks at parameters that are not used here.

Here is a couple of ideas:
```python
        pp_in = (self._ppsur, self._ppa0, self._ppa1)        
        if not pp_in.count(self.pp_to_be_computed):
            return pp_in

        # Does this belong to this method?
        if self._ldbletanh and self._ppa2 * self._ppkth2 * self._ppacr2 == 0.0:
            raise ValueError(
                "ppa2, ppkth2 and ppacr2 must > 0. " "when ldbletanh = True"
            )

        ppacr = self._ppacr
        ppkth = self._ppkth
        
        aa = self._ppdzmin - self._pphmax / (self._jpk - 1)
        bb = np.tanh((1 - ppkth) / ppacr)
        cc = ppacr / (self._jpk - 1)
        dd = np.log(np.cosh((self._jpk - ppkth) / ppacr))
        dd -= np.log(np.cosh((1 - ppkth) / ppacr))

        ppa1 = aa / (bb - cc * dd)
        ppa0 = self._ppdzmin - ppa1 * bb
        ppsur = -(ppa0 + ppa1 * ppacr * np.log(np.cosh((1 - ppkth) / ppacr)))
        
        return (ppsur, ppa0, ppa1)
```
Then, outside of this method:
```python
self._ppsur, self._ppa0, self._ppa1 = self._compute_pp()
```

Reply:
And we can probably introduce a new variable used twice: `np.log(np.cosh((1 - ppkth) / ppacr))`

Reply:
I really like the first idea - I am going to implement it now. Regarding the second comment, I think I already addressed t it - see https://github.com/pyNEMO/pyDOMCFG/pull/15/commits/acf253a4422292eec1081d4a2f05d5a065f14c88 and https://github.com/pyNEMO/pyDOMCFG/pull/15/commits/d0194b486d5173fe147e26bd96a6dc787ef5c179 commits, what do you think?

Reply:
> Regarding the second comment, I think I already addressed it

Do you mean this comment?  https://github.com/pyNEMO/pyDOMCFG/pull/15#discussion_r647224667
I just meant to introduce `ee` within this method.

Reply:
Yes. I agree with you it is redundant to have two methods doing the same thing. 
This methods create a stretched version of the z-coordinate: you can have a one log/cosh function or two. I order to be consistent with the idea that z-ccord are just a special case of s-coordinates, I reather prefer to keep it as methods than a variable. However, in my two last commit I wrote one sinlge genera method.

Reply:
Mmm OK. Not sure if we are talking about the same thing, but I think the way you have it set up right now looks OK to me.
The comment was just about splitting within `compute_pp`:
```python
dd = np.log(np.cosh((self._jpk - self._ppkth) / self._ppacr)) - np.log(np.cosh((1 - self._ppkth) / self._ppacr))
```
in
```python
dd = np.log(np.cosh((self._jpk - ppkth) / ppacr))
ee = np.log(np.cosh((1 - ppkth) / ppacr))
```
to make it easier to follow (`ee` can be used twice). Up to you though!

Reply:
AHHH, I see now ... sorry I didn't understand you, yes we were speaking about two different things :) :) fine by me, with ee it is definetely cleaner ;)